{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "adb587e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import qiime2\n",
    "import pandas as pd\n",
    "import os\n",
    "from qiime2.plugins import (feature_table, alignment, phylogeny, diversity)\n",
    "from tqdm import tqdm\n",
    "\n",
    "base_dir = os.getcwd()\n",
    "results_dir = os.path.join(base_dir, 'results')\n",
    "merged_dir = os.path.join(results_dir, 'merged')\n",
    "data_dir = os.path.join(base_dir, 'data')\n",
    "\n",
    "# merged_alpha_dir: Directory to save alpha diversity results\n",
    "merged_alpha_dir = os.path.join(merged_dir, 'alpha')\n",
    "\n",
    "def cal_alpha_diversity(sample_metadata_df, table_clean, rooted_tree):\n",
    "    '''Calculate alpha diversity metrics and return a DataFrame with results.\n",
    "    Args:\n",
    "        sample_metadata_df (pd.DataFrame): DataFrame containing sample metadata.\n",
    "        table_clean (qiime2.Artifact): Cleaned feature table artifact.\n",
    "        rooted_tree (qiime2.Artifact): Rooted phylogenetic tree artifact.\n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame containing alpha diversity metrics.'''\n",
    "    alpha_shannon = diversity.pipelines.alpha(table = table_clean, \n",
    "                                              metric = 'shannon').alpha_diversity.view(qiime2.Metadata).to_dataframe()\n",
    "    alpha_evenness = diversity.pipelines.alpha(table = table_clean, \n",
    "                                                     metric = 'pielou_e').alpha_diversity.view(qiime2.Metadata).to_dataframe()\n",
    "    alpha_observed_features = diversity.pipelines.alpha(table = table_clean, \n",
    "                                                        metric = 'observed_features').alpha_diversity.view(qiime2.Metadata).to_dataframe()\n",
    "    alpha_faith_pd = diversity.pipelines.alpha_phylogenetic(table = table_clean, \n",
    "                                                            phylogeny = rooted_tree, \n",
    "                                                            metric='faith_pd').alpha_diversity.view(qiime2.Metadata).to_dataframe()\n",
    "\n",
    "    alpha_df = pd.merge(sample_metadata_df[['ProjectID','HostName','BodySite', 'BodysiteClass']],\n",
    "                        alpha_shannon, \n",
    "                        left_index=True, right_index=True, how='left')\n",
    "    alpha_df = pd.merge(alpha_df, \n",
    "                        alpha_evenness, left_index=True, right_index=True, how='left')\n",
    "    alpha_df = pd.merge(alpha_df, \n",
    "                        alpha_observed_features, left_index=True, right_index=True, how='left')\n",
    "    alpha_df = pd.merge(alpha_df, \n",
    "                        alpha_faith_pd, left_index=True, right_index=True, how='left')\n",
    "    return alpha_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e3a5509c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: mafft --preservecase --inputorder --thread 16 /tmp/qiime2/gaoyuze/data/35829b82-f43b-4bb9-9067-8a82217f6a33/data/dna-sequences.fasta\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "inputfile = orig\n",
      "1131 x 390 - 145 d\n",
      "nthread = 16\n",
      "nthreadpair = 16\n",
      "nthreadtb = 16\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      " 1101 / 1131 (thread   12)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      " 1120 / 1131\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   103 / 1130 (thread   11)\n",
      "Reallocating..done. *alloclen = 1781\n",
      "STEP  1101 / 1130 (thread    7)\n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      " 1100 / 1131 (thread    3)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      " 1120 / 1131\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   382 / 1130 (thread   13)\n",
      "Reallocating..done. *alloclen = 1782\n",
      "STEP  1101 / 1130 (thread    7) h\n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.520\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "16 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-2 (Fast but rough)\n",
      " Progressive method (guide trees were built 2 times.)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: FastTreeMP -quote -nt /tmp/qiime2/gaoyuze/data/f52b7e00-570b-4fe2-89f9-9b3de90c9d0a/data/aligned-dna-sequences.fasta\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FastTree Version 2.1.11 Double precision (No SSE3), OpenMP (16 threads)\n",
      "Alignment: /tmp/qiime2/gaoyuze/data/f52b7e00-570b-4fe2-89f9-9b3de90c9d0a/data/aligned-dna-sequences.fasta\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: SH-like 1000\n",
      "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
      "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
      "ML Model: Jukes-Cantor, CAT approximation with 20 rate categories\n",
      "      0.13 seconds: Joined    200 of   1113\n",
      "      0.23 seconds: Joined    400 of   1113\n",
      "      0.34 seconds: Joined    600 of   1113\n",
      "      0.44 seconds: Joined    800 of   1113\n",
      "      0.54 seconds: Joined   1000 of   1113\n",
      "Initial topology in 0.62 seconds\n",
      "Refining topology: 40 rounds ME-NNIs, 2 rounds ME-SPRs, 20 rounds ML-NNIs\n",
      "      0.65 seconds: ME NNI round 2 of 40, 101 of 1114 splits, 12 changes (max delta 0.015)\n",
      "      0.80 seconds: SPR round   1 of   2, 101 of 2230 nodes\n",
      "      0.93 seconds: SPR round   1 of   2, 401 of 2230 nodes\n",
      "      1.03 seconds: SPR round   1 of   2, 601 of 2230 nodes\n",
      "      1.14 seconds: SPR round   1 of   2, 801 of 2230 nodes\n",
      "      1.27 seconds: SPR round   1 of   2, 1101 of 2230 nodes\n",
      "      1.41 seconds: SPR round   1 of   2, 1401 of 2230 nodes\n",
      "      1.54 seconds: SPR round   1 of   2, 1701 of 2230 nodes\n",
      "      1.66 seconds: SPR round   1 of   2, 1901 of 2230 nodes\n",
      "      1.76 seconds: SPR round   1 of   2, 2101 of 2230 nodes\n",
      "      1.86 seconds: ME NNI round 14 of 40, 901 of 1114 splits, 16 changes (max delta 0.016)\n",
      "      2.01 seconds: SPR round   2 of   2, 201 of 2230 nodes\n",
      "      2.16 seconds: SPR round   2 of   2, 501 of 2230 nodes\n",
      "      2.31 seconds: SPR round   2 of   2, 801 of 2230 nodes\n",
      "      2.44 seconds: SPR round   2 of   2, 1101 of 2230 nodes\n",
      "      2.58 seconds: SPR round   2 of   2, 1401 of 2230 nodes\n",
      "      2.70 seconds: SPR round   2 of   2, 1601 of 2230 nodes\n",
      "      2.83 seconds: SPR round   2 of   2, 1901 of 2230 nodes\n",
      "      2.97 seconds: SPR round   2 of   2, 2201 of 2230 nodes\n",
      "Total branch-length 39.319 after 3.08 sec\n",
      "      3.08 seconds: ML Lengths 1 of 1114 splits\n",
      "      3.19 seconds: ML Lengths 401 of 1114 splits\n",
      "      3.29 seconds: ML Lengths 801 of 1114 splits\n",
      "      3.42 seconds: ML NNI round 1 of 20, 101 of 1114 splits, 15 changes (max delta 2.221)\n",
      "      3.56 seconds: ML NNI round 1 of 20, 401 of 1114 splits, 76 changes (max delta 7.138)\n",
      "      3.71 seconds: ML NNI round 1 of 20, 701 of 1114 splits, 137 changes (max delta 9.441)\n",
      "      3.81 seconds: ML NNI round 1 of 20, 901 of 1114 splits, 184 changes (max delta 10.111)\n",
      "ML-NNI round 1: LogLk = -45864.503 NNIs 221 max delta 10.11 Time 3.93\n",
      "      3.94 seconds: Site likelihoods with rate category 1 of 20\n",
      "      4.05 seconds: Site likelihoods with rate category 10 of 20\n",
      "      4.15 seconds: Site likelihoods with rate category 19 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.947 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
      "      4.28 seconds: ML NNI round 2 of 20, 201 of 1114 splits, 19 changes (max delta 3.357)\n",
      "      4.39 seconds: ML NNI round 2 of 20, 401 of 1114 splits, 45 changes (max delta 6.797)\n",
      "      4.50 seconds: ML NNI round 2 of 20, 601 of 1114 splits, 64 changes (max delta 6.797)\n",
      "      4.61 seconds: ML NNI round 2 of 20, 801 of 1114 splits, 87 changes (max delta 6.797)\n",
      "      4.72 seconds: ML NNI round 2 of 20, 1001 of 1114 splits, 115 changes (max delta 7.679)\n",
      "ML-NNI round 2: LogLk = -38298.614 NNIs 125 max delta 7.68 Time 4.79\n",
      "      4.84 seconds: ML NNI round 3 of 20, 101 of 1114 splits, 10 changes (max delta 4.794)\n",
      "      4.95 seconds: ML NNI round 3 of 20, 301 of 1114 splits, 36 changes (max delta 6.288)\n",
      "      5.05 seconds: ML NNI round 3 of 20, 501 of 1114 splits, 45 changes (max delta 6.288)\n",
      "      5.16 seconds: ML NNI round 3 of 20, 701 of 1114 splits, 60 changes (max delta 6.288)\n",
      "ML-NNI round 3: LogLk = -38222.253 NNIs 65 max delta 6.29 Time 5.22\n",
      "      5.27 seconds: ML NNI round 4 of 20, 101 of 1114 splits, 8 changes (max delta 0.961)\n",
      "      5.38 seconds: ML NNI round 4 of 20, 301 of 1114 splits, 24 changes (max delta 7.105)\n",
      "      5.49 seconds: ML NNI round 4 of 20, 501 of 1114 splits, 37 changes (max delta 7.105)\n",
      "ML-NNI round 4: LogLk = -38195.653 NNIs 37 max delta 7.10 Time 5.50\n",
      "      5.61 seconds: ML NNI round 5 of 20, 201 of 1114 splits, 8 changes (max delta 0.971)\n",
      "ML-NNI round 5: LogLk = -38185.816 NNIs 12 max delta 2.84 Time 5.69\n",
      "      5.74 seconds: ML NNI round 6 of 20, 101 of 1114 splits, 6 changes (max delta 1.748)\n",
      "ML-NNI round 6: LogLk = -38179.999 NNIs 6 max delta 1.75 Time 5.80\n",
      "ML-NNI round 7: LogLk = -38177.757 NNIs 3 max delta 1.50 Time 5.84\n",
      "ML-NNI round 8: LogLk = -38177.669 NNIs 3 max delta 0.03 Time 5.89\n",
      "Turning off heuristics for final round of ML NNIs (converged)\n",
      "      5.88 seconds: ML NNI round 9 of 20, 1 of 1114 splits\n",
      "      5.99 seconds: ML NNI round 9 of 20, 201 of 1114 splits, 5 changes (max delta 0.243)\n",
      "      6.09 seconds: ML NNI round 9 of 20, 401 of 1114 splits, 10 changes (max delta 0.807)\n",
      "      6.21 seconds: ML NNI round 9 of 20, 601 of 1114 splits, 11 changes (max delta 0.807)\n",
      "      6.32 seconds: ML NNI round 9 of 20, 801 of 1114 splits, 13 changes (max delta 0.807)\n",
      "      6.42 seconds: ML NNI round 9 of 20, 1001 of 1114 splits, 15 changes (max delta 0.807)\n",
      "ML-NNI round 9: LogLk = -38170.350 NNIs 16 max delta 0.81 Time 6.49 (final)\n",
      "      6.53 seconds: ML Lengths 201 of 1114 splits\n",
      "      6.63 seconds: ML Lengths 701 of 1114 splits\n",
      "Optimize all lengths: LogLk = -38169.545 Time 6.73\n",
      "      6.82 seconds: ML split tests for    100 of   1113 internal splits\n",
      "      6.93 seconds: ML split tests for    200 of   1113 internal splits\n",
      "      7.03 seconds: ML split tests for    300 of   1113 internal splits\n",
      "      7.23 seconds: ML split tests for    500 of   1113 internal splits\n",
      "      7.33 seconds: ML split tests for    600 of   1113 internal splits\n",
      "      7.44 seconds: ML split tests for    700 of   1113 internal splits\n",
      "      7.64 seconds: ML split tests for    900 of   1113 internal splits\n",
      "      7.74 seconds: ML split tests for   1000 of   1113 internal splits\n",
      "Total time: 7.86 seconds Unique: 1116/1131 Bad splits: 1/1113 Worst delta-LogLk 4.789\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/home/gaoyuze/symbiotic_microbiota_analysis/results/merged/rooted_tree.qza'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Constructs a phylogenetic tree.\n",
    "merged_rep_seqs = qiime2.Artifact.load(os.path.join(merged_dir, 'merged_rep_seqs.qza'))\n",
    "aligned_rep_seqs = alignment.methods.mafft(merged_rep_seqs, n_threads = 16).alignment\n",
    "masked_aligned_rep_seqs = alignment.methods.mask(aligned_rep_seqs).masked_alignment\n",
    "masked_aligned_rep_seqs.save(os.path.join(merged_dir, 'masked_aligned_rep_seqs'))\n",
    "masked_aligned_rep_seqs = qiime2.Artifact.load(os.path.join(merged_dir, 'masked_aligned_rep_seqs.qza'))\n",
    "unrooted_tree = phylogeny.methods.fasttree(masked_aligned_rep_seqs, n_threads = 16).tree\n",
    "unrooted_tree.save(os.path.join(merged_dir, 'unrooted_tree'))\n",
    "rooted_tree = phylogeny.methods.midpoint_root(unrooted_tree).rooted_tree\n",
    "rooted_tree.save(os.path.join(merged_dir, 'rooted_tree'))\n",
    "# rooted_tree = qiime2.Artifact.load(os.path.join(merged_dir, 'rooted_tree.qza'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2cffb67c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command:\n",
      "\n",
      "faithpd -i /tmp/qiime2/gaoyuze/data/0036ba6b-337e-4a6e-b5f4-0309d2b4dc06/data/feature-table.biom -t /tmp/qiime2/gaoyuze/data/73382eed-be36-45ce-b307-2c12fafee5a4/data/tree.nwk -o /tmp/q2-AlphaDiversityFormat-a2h0g6u7\n",
      "\n",
      "STEP 5  Done!\n"
     ]
    }
   ],
   "source": [
    "sample_metadata = qiime2.Metadata.load(os.path.join(data_dir, 'sample_metadata.txt'))\n",
    "merged_table_clean = qiime2.Artifact.load(os.path.join(merged_dir, 'merged_table_clean.qza'))\n",
    "\n",
    "alpha_df = pd.DataFrame(\n",
    "    columns=['SampleID','ProjectID', 'HostName', 'BodySite','shannon_entropy','pielou_evenness', 'observed_features', 'faith_pd']\n",
    ")\n",
    "alpha_df = cal_alpha_diversity(sample_metadata.to_dataframe(), merged_table_clean, rooted_tree)\n",
    "alpha_df.reset_index(inplace=True)\n",
    "alpha_df.to_csv(os.path.join(merged_alpha_dir, 'alpha_diversity.csv'), index=False)\n",
    "\n",
    "print('STEP 5  Done!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c22f71c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>ProjectID</th>\n",
       "      <th>HostName</th>\n",
       "      <th>BodySite</th>\n",
       "      <th>BodysiteClass</th>\n",
       "      <th>shannon_entropy</th>\n",
       "      <th>pielou_evenness</th>\n",
       "      <th>observed_features</th>\n",
       "      <th>faith_pd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR23216662</td>\n",
       "      <td>PRJNA924021</td>\n",
       "      <td>Terrapene carolina</td>\n",
       "      <td>Cloaca</td>\n",
       "      <td>CloacaReptilia</td>\n",
       "      <td>4.461310</td>\n",
       "      <td>0.618814</td>\n",
       "      <td>148.0</td>\n",
       "      <td>10.186701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR23216663</td>\n",
       "      <td>PRJNA924021</td>\n",
       "      <td>Terrapene carolina</td>\n",
       "      <td>Cloaca</td>\n",
       "      <td>CloacaReptilia</td>\n",
       "      <td>4.299742</td>\n",
       "      <td>0.608504</td>\n",
       "      <td>134.0</td>\n",
       "      <td>9.457109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR23216664</td>\n",
       "      <td>PRJNA924021</td>\n",
       "      <td>Terrapene carolina</td>\n",
       "      <td>Cloaca</td>\n",
       "      <td>CloacaReptilia</td>\n",
       "      <td>4.197208</td>\n",
       "      <td>0.610915</td>\n",
       "      <td>117.0</td>\n",
       "      <td>8.937190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR23216665</td>\n",
       "      <td>PRJNA924021</td>\n",
       "      <td>Terrapene carolina</td>\n",
       "      <td>Cloaca</td>\n",
       "      <td>CloacaReptilia</td>\n",
       "      <td>4.419209</td>\n",
       "      <td>0.618093</td>\n",
       "      <td>142.0</td>\n",
       "      <td>9.666447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR23216666</td>\n",
       "      <td>PRJNA924021</td>\n",
       "      <td>Terrapene carolina</td>\n",
       "      <td>Cloaca</td>\n",
       "      <td>CloacaReptilia</td>\n",
       "      <td>4.248253</td>\n",
       "      <td>0.588468</td>\n",
       "      <td>149.0</td>\n",
       "      <td>10.929740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR23216667</td>\n",
       "      <td>PRJNA924021</td>\n",
       "      <td>Terrapene carolina</td>\n",
       "      <td>Cloaca</td>\n",
       "      <td>CloacaReptilia</td>\n",
       "      <td>4.695012</td>\n",
       "      <td>0.650354</td>\n",
       "      <td>149.0</td>\n",
       "      <td>10.929740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ERR1474235</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>feces</td>\n",
       "      <td>FecesReptilia</td>\n",
       "      <td>5.288433</td>\n",
       "      <td>0.717920</td>\n",
       "      <td>165.0</td>\n",
       "      <td>11.020333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERR1474236</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>oral cavity</td>\n",
       "      <td>Oral_cavityReptilia</td>\n",
       "      <td>6.785875</td>\n",
       "      <td>0.779326</td>\n",
       "      <td>418.0</td>\n",
       "      <td>24.741341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ERR1474237</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>skin</td>\n",
       "      <td>SkinReptilia</td>\n",
       "      <td>5.491313</td>\n",
       "      <td>0.780745</td>\n",
       "      <td>131.0</td>\n",
       "      <td>10.852530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ERR1474238</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>feces</td>\n",
       "      <td>FecesReptilia</td>\n",
       "      <td>4.663354</td>\n",
       "      <td>0.746384</td>\n",
       "      <td>76.0</td>\n",
       "      <td>5.798360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ERR1474239</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>oral cavity</td>\n",
       "      <td>Oral_cavityReptilia</td>\n",
       "      <td>5.960530</td>\n",
       "      <td>0.774750</td>\n",
       "      <td>207.0</td>\n",
       "      <td>13.912648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ERR1474240</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>feces</td>\n",
       "      <td>FecesReptilia</td>\n",
       "      <td>3.506722</td>\n",
       "      <td>0.537547</td>\n",
       "      <td>92.0</td>\n",
       "      <td>7.823783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ERR1474244</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>skin</td>\n",
       "      <td>SkinReptilia</td>\n",
       "      <td>4.734163</td>\n",
       "      <td>0.638941</td>\n",
       "      <td>170.0</td>\n",
       "      <td>11.930946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ERR1474246</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>oral cavity</td>\n",
       "      <td>Oral_cavityReptilia</td>\n",
       "      <td>4.844038</td>\n",
       "      <td>0.585303</td>\n",
       "      <td>310.0</td>\n",
       "      <td>19.986006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ERR1474247</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>skin</td>\n",
       "      <td>SkinReptilia</td>\n",
       "      <td>4.182698</td>\n",
       "      <td>0.520662</td>\n",
       "      <td>262.0</td>\n",
       "      <td>17.202763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ERR1474248</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>feces</td>\n",
       "      <td>FecesReptilia</td>\n",
       "      <td>4.777043</td>\n",
       "      <td>0.648498</td>\n",
       "      <td>165.0</td>\n",
       "      <td>10.162000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ERR1474249</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>oral cavity</td>\n",
       "      <td>Oral_cavityReptilia</td>\n",
       "      <td>4.741792</td>\n",
       "      <td>0.728633</td>\n",
       "      <td>91.0</td>\n",
       "      <td>7.808734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ERR1474250</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>skin</td>\n",
       "      <td>SkinReptilia</td>\n",
       "      <td>6.209409</td>\n",
       "      <td>0.827938</td>\n",
       "      <td>181.0</td>\n",
       "      <td>12.044051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ERR1474251</td>\n",
       "      <td>PRJEB14602</td>\n",
       "      <td>Varanus komodoensis</td>\n",
       "      <td>feces</td>\n",
       "      <td>FecesReptilia</td>\n",
       "      <td>4.595366</td>\n",
       "      <td>0.625328</td>\n",
       "      <td>163.0</td>\n",
       "      <td>11.084091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SampleID    ProjectID             HostName     BodySite  \\\n",
       "0   SRR23216662  PRJNA924021   Terrapene carolina       Cloaca   \n",
       "1   SRR23216663  PRJNA924021   Terrapene carolina       Cloaca   \n",
       "2   SRR23216664  PRJNA924021   Terrapene carolina       Cloaca   \n",
       "3   SRR23216665  PRJNA924021   Terrapene carolina       Cloaca   \n",
       "4   SRR23216666  PRJNA924021   Terrapene carolina       Cloaca   \n",
       "5   SRR23216667  PRJNA924021   Terrapene carolina       Cloaca   \n",
       "6    ERR1474235   PRJEB14602  Varanus komodoensis        feces   \n",
       "7    ERR1474236   PRJEB14602  Varanus komodoensis  oral cavity   \n",
       "8    ERR1474237   PRJEB14602  Varanus komodoensis         skin   \n",
       "9    ERR1474238   PRJEB14602  Varanus komodoensis        feces   \n",
       "10   ERR1474239   PRJEB14602  Varanus komodoensis  oral cavity   \n",
       "11   ERR1474240   PRJEB14602  Varanus komodoensis        feces   \n",
       "12   ERR1474244   PRJEB14602  Varanus komodoensis         skin   \n",
       "13   ERR1474246   PRJEB14602  Varanus komodoensis  oral cavity   \n",
       "14   ERR1474247   PRJEB14602  Varanus komodoensis         skin   \n",
       "15   ERR1474248   PRJEB14602  Varanus komodoensis        feces   \n",
       "16   ERR1474249   PRJEB14602  Varanus komodoensis  oral cavity   \n",
       "17   ERR1474250   PRJEB14602  Varanus komodoensis         skin   \n",
       "18   ERR1474251   PRJEB14602  Varanus komodoensis        feces   \n",
       "\n",
       "          BodysiteClass  shannon_entropy  pielou_evenness  observed_features  \\\n",
       "0        CloacaReptilia         4.461310         0.618814              148.0   \n",
       "1        CloacaReptilia         4.299742         0.608504              134.0   \n",
       "2        CloacaReptilia         4.197208         0.610915              117.0   \n",
       "3        CloacaReptilia         4.419209         0.618093              142.0   \n",
       "4        CloacaReptilia         4.248253         0.588468              149.0   \n",
       "5        CloacaReptilia         4.695012         0.650354              149.0   \n",
       "6         FecesReptilia         5.288433         0.717920              165.0   \n",
       "7   Oral_cavityReptilia         6.785875         0.779326              418.0   \n",
       "8          SkinReptilia         5.491313         0.780745              131.0   \n",
       "9         FecesReptilia         4.663354         0.746384               76.0   \n",
       "10  Oral_cavityReptilia         5.960530         0.774750              207.0   \n",
       "11        FecesReptilia         3.506722         0.537547               92.0   \n",
       "12         SkinReptilia         4.734163         0.638941              170.0   \n",
       "13  Oral_cavityReptilia         4.844038         0.585303              310.0   \n",
       "14         SkinReptilia         4.182698         0.520662              262.0   \n",
       "15        FecesReptilia         4.777043         0.648498              165.0   \n",
       "16  Oral_cavityReptilia         4.741792         0.728633               91.0   \n",
       "17         SkinReptilia         6.209409         0.827938              181.0   \n",
       "18        FecesReptilia         4.595366         0.625328              163.0   \n",
       "\n",
       "     faith_pd  \n",
       "0   10.186701  \n",
       "1    9.457109  \n",
       "2    8.937190  \n",
       "3    9.666447  \n",
       "4   10.929740  \n",
       "5   10.929740  \n",
       "6   11.020333  \n",
       "7   24.741341  \n",
       "8   10.852530  \n",
       "9    5.798360  \n",
       "10  13.912648  \n",
       "11   7.823783  \n",
       "12  11.930946  \n",
       "13  19.986006  \n",
       "14  17.202763  \n",
       "15  10.162000  \n",
       "16   7.808734  \n",
       "17  12.044051  \n",
       "18  11.084091  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alpha_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-2023.7",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
